Add adapter for C assuming Meson build system and Criterion unit testing framework

oyvindaakre opened the pull request:
Add an adapter for unit tests written in C using the [Criterion](https://github.com/Snaipe/Criterion) unit test framework and assuming the [meson](https://mesonbuild.com/) build system.

Note: Build directory assumptions:
- Defaults to `build` unless set by the user when configuring the adapter
- build directory has been set up and configured by a previous call to `meson setup build` (the adapter will automatically build the project before each test).

### Dependencies
* `meson` installed on the system
* `ninja` installed on the system

### Supported features
* Run test closest to the cursor
* Run all tests in the file

quolpr commented:
@oyvindaakre looks great! Ready for merge?

quolpr commented:
@oyvindaakre also, could add please example to test dir, like I did for go - `test/gotest`. The implementation should be very simple, just text for like 1 + 1 == 2, or as you want

quolpr commented:
@quolpr btw, you reload plugin while developing with this command:

```
vim.api.nvim_create_user_command("C", function()
  require("lazy.core.loader").reload("quicktest.nvim")
end)
```

So, what I usually do:
1. Add sample project to test folder
2. Edit adapter
3. Run `:C` command
4. Rerun test

quolpr commented:
@oyvindaakre

I made some breaking changes in API. Changes are not so much, please check Readme 🙂

oyvindaakre commented:
@quolpr Should this PR update the `doc` too?

quolpr commented:
@oyvindaakre the doc will be generated from Readme once PR is merged to main. So, please, mention your adapter in Readme 🙂

oyvindaakre commented:
@quolpr Finally, I think the adapter is ready. I have two questions though:
1) The adapter needs to know the path to the build directory. Currently I hardcode `build` here: https://github.com/oyvindaakre/quicktest.nvim/blob/ea9396b6660116e220f575a5aa33743fe8a1a13a/lua/quicktest/adapters/criterion/init.lua#L12. Is it possible for users to override this and set a custom path if they want? 

2) I added a sample project to `tests/support/criterion`. I also added a readme in an attempt to explain how to set up the project (+ some notes on the adapter itself). 
I'm not sure how to run it with the makefile. When I run `make test` it only seems to run the plugin_name_spec.lua test. Is that expected?
```bash
$ make test
Starting...	
Scheduling: tests/plugin_name/plugin_name_spec.lua

========================================	
Testing: 	/home/oyvind/.local/share/nvim/lazy/quicktest.nvim/tests/plugin_name/plugin_name_spec.lua	
Success	||	setup works with default	
	
Success: 	1	
Failed : 	0	
Errors : 	0	
========================================	
```

quolpr commented:
@oyvindaakre as for
> The adapter needs to know the path to the build directory. Currently I hardcode build here: https://github.com/oyvindaakre/quicktest.nvim/blob/ea9396b6660116e220f575a5aa33743fe8a1a13a/lua/quicktest/adapters/criterion/init.lua#L12. Is it possible for users to override this and set a custom path if they want?

Yep, I think the best way will be to make your plugin to work like this:
```
qt.setup({
  adapters = {
    require("quicktest.adapters.golang")({
       exec_path = function(params)
       end
    }),
  }
})
```

You can do this with meta table, check example here
https://github.com/fredrikaverpil/neotest-golang/blob/main/lua/neotest-golang/init.lua#L179

I will be adding the same API to other adapters too soon. Also, notice that function is supplied for param, that's how user will be able to change behaviour per project. Based on what `vim.fn.bufname(params.bufnr)`(returns buffer path) contains, they will be able to dynamically configure exec_path.

Let me know what you think

> I'm not sure how to run it with the makefile. When I run make test it only seems to run the plugin_name_spec.lua test. Is that expected?

Ha-ha, I didn't make tests for quicktet 😄 So only manual tests for now

quolpr commented on lua/quicktest/adapters/criterion/init.lua:
```diff
@@ -0,0 +1,166 @@
+  local line = criterion.get_nearest_test(bufnr, cursor_pos)
+
+  if line == "" or test_exe == "" then
+    return {}
```
You can return the second params wit error, like return {}, "Failed to find exe". The change is already in master, you need merge it

oyvindaakre replied:
Please see 420f22964b1206951c11fcf1e2d8b7742593686b

quolpr commented on lua/quicktest/adapters/criterion/init.lua:
```diff
@@ -0,0 +1,166 @@
+--- Attempt to run if params is not an empty table
+---@param params CriterionTestParams
+---@return boolean
+M.can_run = function(params)
```
Not needed any more, cause build_* functions return error as second return value

oyvindaakre commented on lua/quicktest.lua:
```diff
@@ -5,6 +5,7 @@ local config = {
   adapters = {
     require("quicktest.adapters.golang"),
     require("quicktest.adapters.elixir"),
+    require("quicktest.adapters.criterion"),
```
Should be fixed by e6e9293fb036539f31af240fa97cb01207036326

oyvindaakre commented:
> Let me know what you think
It sounds neat to support both function and string, but I don't have enough Lua-experience under my belt to know the implications of supporting both :) Metatable is a new concept for me, so I have to read up on it, but I'm happy to take suggestions on how to implement it for this adapter.

quolpr commented:
@oyvindaakre I made config options support for golang, could you do the same way? https://github.com/quolpr/quicktest.nvim/commit/a3bb0b6bded24ba2141cda023bf9161383ac4e30

oyvindaakre commented:
> @oyvindaakre I made config options support for golang, could you do the same way? [a3bb0b6](https://github.com/quolpr/quicktest.nvim/commit/a3bb0b6bded24ba2141cda023bf9161383ac4e30)

Looks good! Copied your steps. Does it look okay?

quolpr commented:
@oyvindaakre LGTM! Thank you for the contribution, merging it now :)
